internal/worker/dialer.go

```diff
@@ -0,0 +1,114 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.uber.org/zap"
+)
+
+// Cache to store resolved IPs
+type hostCache struct {
+	mu    sync.RWMutex
+	cache map[string]string // hostname -> IP address
+}
+
+func newHostCache() *hostCache {
+	return &hostCache{
+		cache: make(map[string]string),
+	}
+}
+
+func (hc *hostCache) Get(hostname string) (string, bool) {
+	hc.mu.RLock()
+	defer hc.mu.RUnlock()
+	ip, ok := hc.cache[hostname]
+	return ip, ok
+}
+
+func (hc *hostCache) Set(hostname, ip string) {
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+	hc.cache[hostname] = ip
+}
+
+func (hc *hostCache) Delete(hostname string) {
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+	delete(hc.cache, hostname)
+}
+
+type DialerBus interface {
+	Host(ctx context.Context, hostKey types.PublicKey) (api.Host, error)
+}
+
+// FallbackDialer implements a custom net.Dialer with a fallback mechanism
+type FallbackDialer struct {
+	cache *hostCache
+
+	bus    DialerBus
+	logger *zap.SugaredLogger
+	dialer net.Dialer
+}
+
+func NewFallbackDialer(bus DialerBus, logger *zap.Logger, dialer net.Dialer) *FallbackDialer {
+	return &FallbackDialer{
+		cache: newHostCache(),
+
+		bus:    bus,
+		logger: logger.Sugar().Named("fallbackdialer"),
+		dialer: dialer,
+	}
+}
+
+func (d *FallbackDialer) Dial(ctx context.Context, hk types.PublicKey, address string) (net.Conn, error) {
+	host, port, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+
+	// Try to resolve IP
+	ipAddr, err := net.ResolveIPAddr("ip", host)
```

Comment:
Can't we skip `ResolveIPAddr` and call `DialContext` right away? Then get the IP address off the `conn`?